### PR TITLE
Initial commit for Concat KDF

### PR DESCRIPTION
**Pull request by rpiasetskyi**
This key derivation function is described in the section 5.8.1 of [1].

[1]: https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-56ar.pdf

Reviewed-by: Norbert Fabritius <n.fabritius@posteo.de>

### CODE REVIEW

**Comment by rpiasetskyi** [comment]
This PR is linked to https://github.com/RustCrypto/KDFs/issues/58.

About failing `workspace /clippy` check:
```
error[E0599]: no associated item named `MAX` found for type `u32` in the current scope
  --> concat-kdf/src/lib.rs:75:75
   |
75 |     if OVERFLOW_IS_POSSIBLE && (key.len() >= D::OutputSize::USIZE * (u32::MAX as usize)) {
   |                                                                           ^^^ associated item not found in `u32`

error: aborting due to previous error
```
The reason is that [u32:MAX](https://doc.rust-lang.org/std/primitive.u32.html#associatedconstant.MAX) is supported since 1.43.0 and the failing check uses 1.41.0.

There are two solutions:
* Update the check version (I would go with this one).
* Define the constant `U32_MAX` in the crate.

**Comment by tarcieri** [comment]
Feel free to bump the Rust version used with clippy

**Comment by rpiasetskyi** [comment]
Could you please suggest the next steps for this PR? It already had a small review [here](https://github.com/rpiasetskyi/KDFs/pull/1), I think it should be ready for the RustCrypto community review. Should I register the crate at [crates.io](https://crates.io/)? Do I need to add something regarding documentation or tests?

Thank you in advance.

File: README.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ Collection of [Key Derivation Functions][KDF] (KDF) written in pure Rust.
 
 | Algorithm | Crate  | Crates.io     | Documentation | MSRV |
 |-----------|--------|:-------------:|:-------------:|:----:|
-| [HKDF]    | [`hkdf`] | [![crates.io](https://img.shields.io/crates/v/hkdf.svg)](https://crates.io/crates/hkdf) | [![Documentation](https://docs.rs/hkdf/badge.svg)](https://docs.rs/hkdf) | ![MSRV 1.41][msrv-1.41] |
+| [HKDF]    | [`hkdf`] | [![crates.io](https://img.shields.io/crates/v/hkdf.svg)](https://crates.io/crates/hkdf) | [![Documentation](https://docs.rs/hkdf/badge.svg)](https://docs.rs/hkdf) | ![MSRV 1.58.1][msrv-1.58.1] |
```
Comment (tarcieri) [inline README.md:11]: It looks like you didn't update this table correctly. You modified HKDF's entry, rather than adding a new one for `concat-kdf`

File: concat-kdf/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+version = "0.1.0"
+authors = ["Ruslan Piasetskyi <ruslan.piasetskyi@gmail.com>"]
+license = "MIT OR Apache-2.0"
+homepage = "https://github.com/RustCrypto/KDFs/"
+repository = "https://github.com/RustCrypto/KDFs/"
+description = "Concatenation Key Derivation Function (Concat KDF)"
+keywords = ["crypto", "concat-kdf", "KDF", "NIST"]
+categories = ["cryptography", "no-std"]
+readme = "README.md"
+edition = "2018"
```
Comment (tarcieri) [inline concat-kdf/Cargo.toml:12]:
```suggestion
edition = "2021"
rust-version = "1.58.1"
```

...although I'm unclear why the MSRV is 1.58.1. Are you actually using features introduced in that version? Otherwise a 2021 edition crate is MSRV 1.56.

File: hkdf/README.md [hunk 3]
```diff
@@ -78,7 +78,7 @@ assert_eq!(okm, expected);
 [docs-image]: https://docs.rs/hkdf/badge.svg
 [docs-link]: https://docs.rs/hkdf/
 [license-image]: https://img.shields.io/badge/license-Apache2.0/MIT-blue.svg
-[rustc-image]: https://img.shields.io/badge/rustc-1.41+-blue.svg
+[rustc-image]: https://img.shields.io/badge/rustc-1.58.1+-blue.svg
```
Comment (tarcieri) [inline hkdf/README.md:81]: Why did you change this MSRV?

**Comment by tarcieri** [comment]
This is simple enough. But I'm confused why you changed the documented MSRV for HKDF, and why that MSRV was picked.

**Comment by tarcieri** [comment]
I'm going to go ahead and merge this and fix up the issues I noted myself
